New Resource: azurerm_frontdoor_rules_engine

heoelri opened the pull request:
This PR introduces a new resource type called `azurerm_frontdoor_rules_engine`. This resource type is used to create Rules Engine configurations in Azure Front Door Service. 

This was raised and discussed here: https://github.com/hashicorp/terraform-provider-azurerm/issues/7455

Here's a sample definition:

```hcl
resource "azurerm_frontdoor_rules_engine" "cors_rules" {
  name                = "corsrules"
  frontdoor_name      = azurerm_frontdoor.example.name
  resource_group_name = azurerm_frontdoor.example.resource_group_name

  rule {
    name = "debug"
    priority = 1

    match_condition {
      match_variable = "RequestMethod"
      operator = "Equal"
      match_value = ["GET", "POST"]
    }

    rule_action {
      response_header_actions {
        header_action_type = "Append"
        header_name        = "X-TEST-HEADER"
        value              = "CORS CMS Rule"
      }
    }
  }

  rule {
    name     = "origin"
    priority = 2

    rule_action {
      request_header_actions {
        header_action_type = "Overwrite"
        header_name        = "Origin"
        value              = "this is an append test"
      }
      response_header_actions {
        header_action_type = "Overwrite"
        header_name        = "Access-Control-Allow-Origin"
        value              = "*"
      }
    }
  }

  depends_on = [
    azurerm_frontdoor.example
  ]
}
```

And here's the current output of `terraform plan`:

![image](https://user-images.githubusercontent.com/17407022/132334159-7fbfa489-dbcd-4aa2-af3e-9f807f97e844.png)

And here's the result of above's definition in Azure Portal:

![image](https://user-images.githubusercontent.com/17407022/132334631-6fdc0e04-7e35-4147-9606-bb52b310523d.png)

What current does not work is associating these rules to an existing Front Door routing rule. There was a workaround documented here: https://github.com/hashicorp/terraform-provider-azurerm/issues/7455#issuecomment-882769364

# TODO
* Figure out if the association can be done via ARM 
* Adding support for `routeConfigurationOverride`
* Adding support for `MatchProcessingBehavior`

katbyte reviewed (changes requested):
Thanks @heoelri - overall this looks good but i have some suggestions for the schema to improve the ux of the tested properties. WDYT?

heoelri commented on internal/services/frontdoor/frontdoor_rules_engine_rule.go:
```diff
@@ -0,0 +1,458 @@
+							Elem: &pluginsdk.Resource{
+								Schema: map[string]*pluginsdk.Schema{
+
+									"match_variable": {
```
Yes, good catch. I'll commit your suggestion and update the docs.

heoelri replied:
@katbyte i think you're right, but in this case is it consistent with the parameters of `azurerm_frontdoor_firewall_policy`.

The `azurerm_frontdoor_firewall_policy` resource is using the same parameter names:

```hcl
match_condition {
  match_variable     = "RemoteAddr"
  operator           = "IPMatch"
  negation_condition = false
  match_values       = ["192.168.1.0/24", "10.0.0.0/24"]
}
```
Do you still suggest to change it or keep it consistent with the other Front Door resource?

katbyte replied:
lets change it, and if you don't mind adding a todo on the other one to make it consistent that would be great! thanks

heoelri replied:
renamed and updated docs

heoelri commented on internal/services/frontdoor/frontdoor_rules_engine_rule.go:
```diff
@@ -0,0 +1,458 @@
+									// 	Type:         pluginsdk.TypeString,
+									// 	Optional:     true,
+									// 	ValidateFunc: validation.StringIsNotEmpty,
+									// },
```
@katbyte this PR contains only the functionality to create, update and delete new rules engine configurations, rules and conditions. There's a small rest left that does not work today. This includes the capability to associate a rules engine configuration to a front door routing rule. This does not work via ARM today and is only possible via a workaround as discussed in #7455 

I've removed everything that was commented out.

heoelri commented on internal/services/frontdoor/frontdoor_rules_engine_rule.go:
```diff
@@ -0,0 +1,458 @@
+										Default:  true,
+									},
+
+									"match_value": {
```
done

katbyte commented:
looks like we have some test failures:
```
------- Stdout: -------
=== RUN   TestAccFrontDoor_RulesEngine
=== PAUSE TestAccFrontDoor_RulesEngine
=== CONT  TestAccFrontDoor_RulesEngine
    testcase.go:88: Step 1/2 error: Error running pre-apply refresh: exit status 1
        
        Error: Unsupported block type
        
          on terraform_plugin_test.tf line 73, in resource "azurerm_frontdoor_rules_engine" "sample_engine_config":
          73:     rule_action {
        
        Blocks of type "rule_action" are not expected here.
        
        Error: Unsupported block type
        
          on terraform_plugin_test.tf line 86, in resource "azurerm_frontdoor_rules_engine" "sample_engine_config":
          86:     rule_action {
        
        Blocks of type "rule_action" are not expected here.
--- FAIL: TestAccFrontDoor_RulesEngine (16.17s)
FAIL
```

heoelri commented:
> looks like we have some test failures:
> 
> ```
> ------- Stdout: -------
> === RUN   TestAccFrontDoor_RulesEngine
> === PAUSE TestAccFrontDoor_RulesEngine
> === CONT  TestAccFrontDoor_RulesEngine
>     testcase.go:88: Step 1/2 error: Error running pre-apply refresh: exit status 1
>         
>         Error: Unsupported block type
>         
>           on terraform_plugin_test.tf line 73, in resource "azurerm_frontdoor_rules_engine" "sample_engine_config":
>           73:     rule_action {
>         
>         Blocks of type "rule_action" are not expected here.
>         
>         Error: Unsupported block type
>         
>           on terraform_plugin_test.tf line 86, in resource "azurerm_frontdoor_rules_engine" "sample_engine_config":
>           86:     rule_action {
>         
>         Blocks of type "rule_action" are not expected here.
> --- FAIL: TestAccFrontDoor_RulesEngine (16.17s)
> FAIL
> ```

Sorry, my bad. I've updated the code and the docs but not the test. I've now also updated the test to reflect the renamed parameters and blocks. 

Here are the results:

```console
make acctests SERVICE='frontdoor' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/frontdoor  -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
[..]
--- PASS: TestAccFrontDoorFirewallPolicy_complete (91.04s)
=== CONT  TestAccFrontDoor_EnableDisableCache
--- PASS: TestAccFrontDoorFirewallPolicy_requiresImport (102.54s)
=== CONT  TestAccFrontDoor_RulesEngine
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_EnabledWithoutConfigurationBlock (144.55s)
=== CONT  TestAccFrontDoor_waf
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_EnabledKeyVaultLatest (144.93s)
=== CONT  TestAccFrontDoor_complete
--- PASS: TestAccFrontDoor_multiplePools (177.74s)
=== CONT  TestAccFrontDoor_requiresImport
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_EnabledKeyVaultLatestMissingAttributes (204.51s)
=== CONT  TestAccFrontDoor_update
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_EnabledFrontdoorExtraAttributes (205.35s)
=== CONT  TestAccFrontDoorFirewallPolicy_update
--- PASS: TestAccFrontDoor_complete (201.89s)
=== CONT  TestAccFrontDoorFirewallPolicy_basic
--- PASS: TestAccFrontDoor_waf (207.82s)
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_DisabledWithConfigurationBlock
--- PASS: TestAccFrontDoorFirewallPolicy_update (160.95s)
=== CONT  TestAccFrontDoor_global
--- PASS: TestAccFrontDoor_requiresImport (212.50s)
=== CONT  TestAccFrontDoor_basic
--- PASS: TestAccFrontDoorFirewallPolicy_basic (77.82s)
=== CONT  TestAccFrontDoorCustomHttpsConfiguration_EnabledKeyVaultMissingAttributes
--- PASS: TestAccFrontDoor_EnableDisableCache (466.72s)
--- PASS: TestAccFrontDoor_update (456.63s)
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_DisabledWithConfigurationBlock (809.01s)
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_EnabledKeyVaultMissingAttributes (1859.13s)
--- PASS: TestAccFrontDoor_global (2015.65s)
--- PASS: TestAccFrontDoorCustomHttpsConfiguration_CustomHttps (2437.22s)
--- PASS: TestAccFrontDoor_RulesEngine (2343.57s)
--- PASS: TestAccFrontDoor_basic (2086.86s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/frontdoor     2477.152s
```

katbyte reviewed (approved):
Thanks @heoelri! LGTM now 🚀

joaomoreno commented:
@heoelri Thanks for this, can't wait to use it!
